xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
       TF_RETURN_IF_ERROR(UpdateDotAndConsumerConfig(inst, stream_id));
       ++stream_id;
     }
+    // If dot's result is accumulated, this means we found a loop with
+    // contracting dim sharded.
+    HloInstruction* partial_dot;
+    if (Match(inst, m::AddAnyOrder(m::Op(),
+                                   m::Dot(&partial_dot, m::Op(), m::Op())))) {
+      partial_accumulations.push_back(partial_dot);
+    }
+  }
+
+  // Transform partial accumulations into a reduction on a contiguous buffer.
+  // Partial accumulations will impact the overlap between dots because the
+  // dot+add pattern will be fused into a single gemm later in gemm rewriter
+  // which adds data dependencies between gemms. Instead we write all
+  // intermediate results into a larger buffer and perform a one-shot reduction.
+  // The high-level transformation is:
+  // shape(x,y) cp   shape(x,y) dot0
+  //          \         /
+  //           \       /
+  //     shape(x,y) add0    shape(x,y) dot1
+  //             \                /
+  //              \              /
+  //             shape(x,y) add1
+  //                     |
+  //        shape(x,y) loop output
+  //
+  // transformed into:
+  // shape(x,y) cp         shape(x,y) dot0    shape(x,y) dot1
+  //    \                     /                   /
+  //     \                   /                   /
+  //     shape(n,x,y) concatenate on first axis, n is the number of partitions
+  //                        |
+  //             shape(n,x,y) loop output
+  //                        |
+  //             shape(x,y) reduction on first axis
```

Comment:
The diagram doesn't match the test case. In the test case, you do:

accumulator += dot(lhs, slice(rhs))

The diagram shows `cp` as an input to `add0`, but this doesn't happen in the test case. Yet the rewrite still occurs in the test case. Can you make the diagram a bit more general? All that is necessary for the rewrite to happen is a dot output being added to an accumulator.

Also, in what cases is a collective-permute directly fed into an add? I thought an add always has a local accumulator as the LHS, but I am not aware of very many cases.

Reply:
This is my mistake, i just put cp as an arbitrary op to illustrate the flow. I will make it more precise.